### PR TITLE
HV-1973 - Bitcoin address validator

### PR DESCRIPTION
**Pull request by jyoshiriro**
Bitcoin address validator.

More translations are still missing. Only english and portuguese have been provided so far.

See https://hibernate.atlassian.net/browse/HV-1973.

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	 */
+	@Override
+	public boolean isValid(CharSequence charSequence, ConstraintValidatorContext context) {
+
+		if ( charSequence == null ) {
+			return true;
+		}
+
+		for ( BitcoinAddressType type : this.addressType ) {
+			Pattern pattern = Pattern.compile( type.getRegex() );
```
Comment (marko-bekhta) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java:71]: Let's have the patterns pre-compiled.

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+				Arrays.stream( BitcoinAddressType.values() )
+						.filter( type -> type != BitcoinAddressType.ANY )
+						.collect( Collectors.toList() ) );
+			return;
+		}
+
+		Collections.addAll( this.addressType, types );
+		this.typesDescription = this.addressType.stream()
+								.map( BitcoinAddressType::getDescription )
+								.collect( Collectors.joining( "; " ) );
```
Comment (marko-bekhta) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java:53]: once the impl details are moved out of the enum here you could still keep them in the enum and maybe also have this prebuilt? e.g.:
```java
public enum BitcoinAddressConstraint {
	P2PKH("Legacy (P2PKH)", "^(1)[a-zA-HJ-NP-Z0-9]{25,61}$"),
	P2SH("Nested SegWit (P2SH)", "^(3)[a-zA-HJ-NP-Z0-9]{33}$"),
	BECH32("Native SegWit (Bech32)", "^(bc1)[a-zA-HJ-NP-Z0-9]{39,59}$"),
	P2WSH("SegWit variant of P2SH (P2WSH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{58}$"),
	P2WPKH("SegWit variant of P2PKH (P2WPKH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{38}$"),
	P2TR("Taproot (P2TR)", "^(bc1p)[a-zA-HJ-NP-Z0-9]{58}$");

	BitcoinAddressConstraint(String description, String pattern) {
		this.pattern = Pattern.compile( pattern );
	}

	private static final EnumMap<BitcoinAddress.Type, Collection<BitcoinAddressConstraint>> constraints;
	static {
		constraints = new EnumMap<>( BitcoinAddress.Type.class );
		constraints.put( BitcoinAddress.Type.ANY, Arrays.asList( BitcoinAddressConstraint.values() ) );
		constraints.put( BitcoinAddress.Type.P2PKH, Collections.singleton( P2PKH ) );
		constraints.put( BitcoinAddress.Type.P2SH, Collections.singleton( P2SH ) );
		constraints.put( BitcoinAddress.Type.BECH32, Collections.singleton( BECH32 ) );
		constraints.put( BitcoinAddress.Type.P2WSH, Collections.singleton( P2WSH ) );
		constraints.put( BitcoinAddress.Type.P2WPKH, Collections.singleton( P2WPKH ) );
		constraints.put( BitcoinAddress.Type.P2TR, Collections.singleton( P2TR ) );

	}
 ... 
}
```
and then in the initialize:
```java
public void initialize(BitcoinAddress bitcoinAddress) {
	Set<BitcoinAddressConstraint> constraints = new HashSet<>();
	for ( BitcoinAddress.Type type : bitcoinAddress.value() ) {
		constraints.addAll( BitcoinAddressConstraint.constraints.get( type ) );
	}
	this.singleType = constraints.size() == 1;
}
```

File: engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+package org.hibernate.validator.constraints;
+
+public enum BitcoinAddressType {
+	ANY("Bitcoin", null),
+	P2PKH("Legacy (P2PKH)", "^(1)[a-zA-HJ-NP-Z0-9]{25,61}$"),
+	P2SH("Nested SegWit (P2SH)", "^(3)[a-zA-HJ-NP-Z0-9]{33}$"),
+	BECH32("Native SegWit (Bech32)", "^(bc1)[a-zA-HJ-NP-Z0-9]{39,59}$"),
+	P2WSH("SegWit variant of P2SH (P2WSH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{58}$"),
+	P2WPKH("SegWit variant of P2PKH (P2WPKH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{38}$"),
+	P2TR("Taproot (P2TR)", "^(bc1p)[a-zA-HJ-NP-Z0-9]{58}$");
```
Comment (marko-bekhta) [inline engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java:16]: BTW I see that this description is used in the validation message. But as they use regular words they may need to be translated as well... So I'd suggest exploring an option of placing the description in the translation file as well and then let EL do the string interpolation.

**Review by marko-bekhta — COMMENTED** [review]
Hey,

thanks for submitting the pull request with the new constraint. This looks interesting. 
I've added some suggestions inline. Please also take a look at this post https://in.relation.to/2018/01/04/adding-new-constraint-to-engine/ to see about the additional parts that should be added for a constraint.

File: engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.constraints;
+
+public enum BitcoinAddressType {
```
[the remark this answers is not in the document]
Comment (jyoshiriro) [reply engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java]: I did almost that... When you have a chance to take a look please :)

File: engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	}
+
+	public String getRegex() {
+		return regex;
+	}
+
+	@Override
+	public String toString() {
+		return this.description;
+	}
```
[the remark this answers is not in the document]
Comment (jyoshiriro) [reply engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java]: I did almost that... When you have a chance to take a look please :)

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java (hunk 1 above)
Comment (jyoshiriro) [reply engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java]: Done! When you have a chance to take a look please :)

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java (hunk 2 above)
Comment (jyoshiriro) [reply engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java]: I did almost that... When you have a chance to take a look please :)

File: engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java (hunk 3 above)
Comment (jyoshiriro) [reply engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java]: Done. When you have a chance to take a look please :)

**Comment by jyoshiriro** [comment]
> Hey,
> 
> thanks for submitting the pull request with the new constraint. This looks interesting. I've added some suggestions inline. Please also take a look at this post https://in.relation.to/2018/01/04/adding-new-constraint-to-engine/ to see about the additional parts that should be added for a constraint.

I believe I fixed everything that was missing :)

**Comment by marko-bekhta** [comment]
Thanks! I've squashed the changes in this patch and removed the dependency on the resource bundle from the validator implementation. I'm going to close this PR in favour of  https://github.com/hibernate/hibernate-validator/pull/1387 
Thanks for all the work you did here!

**Pull request closed without merging**